### DISCUSSION TITLE
Caching validation result per object

### DISCUSSION THREAD

**Discussion by diznq** [pydantic/pydantic#4322] [Feature Request]
Let there be situation with two Pydantic models

```python
class B:
 number: int

class A:
 numbers: List[B]
``` 

Right now, even if we have a valid list of `B`, when this list is assigned to an existing instance of `A.numbers`, it's revalidated. This becomes issue at larger objects, when such operation is quite CPU intensive and can take more than 10 seconds to complete.

Is it possible to create some sort of `in_tact: bool` meta attribute that is refreshed when assigning primitive types? 
This way if we assigned list of `B` that had `in_tact` set to `True`, since we know it's in tact, we don't need to run costly validation again and we can directly assign it.

**Comment by abi-jey** [comment]
This could be a productivity improvement and would mean that developer does not need to think about if the input is will be as json, dict or already a with same required pydantic model. Can this somehow be planned for v2?

**Reply by samuelcolvin** [reply]
"Validating `b`" is as simple as confirming it's an instance of `B`, so there's no costly validation.

This is improved in V2 where we have [`revalidate_models`](https://github.com/pydantic/pydantic-core/blob/fdf42b8b0d0684bc2cac04d34c40fe4b0d33498d/pydantic_core/_types.py#L39), and also the check is `type(v) == B` rather than `isinstance(b, B)`.

**Reply by diznq** [reply]
Hi, thanks for the response. I tried to profile the code once again and you are right. 

I figured out that the real issue was actually Python's ref-counted garbage collection, which removes objects from memory as soon as possible when they reach zero references. This was a big issue when there were hundreds of millions of objects down the hierarchy and therefore when we did `a.numbers = b.numbers` on top level, garbage collector had to go through all millions of objects of `a` and release them, which took tens of seconds. It's confirmed by the fact that doing `temp = a.numbers; a.numbers=b.numbers` takes just fraction of second.

I'm sorry for wrongly assuming it was on Pydantic's side 😄
